MetaMask/metamask-mobile
# feat: New mobile Flask Build Type

## **Description**

This change enables a new build type for MetaMask mobile. This new build is called `flask` and will be used for the canary distribution of features, namely snaps. These builds will be distributed through the [google play store as a new app listing](https://play.google.com/store/apps/details?id=io.metamask.flask&hl=en&gl=US&pli=1) on android, and through tetsflight on iOS. These builds already exist through the work of [mobile snaps](https://github.com/MetaMask/metamask-mobile/pull/7526/files#diff-70fd746326e47bc54d66a46e360af92a5d36328528897869ba9d97c4877ed62b). This PR pulls in the `non snaps` related code for creating these builds. The end result is new "flavour" of metamask, that is identical to the main build. Once we merge [code fencing](https://github.com/MetaMask/metamask-mobile/pull/7734) into main, we can then start merging snaps code into then flask build without effecting the main build. 

- Most of this work was already implemented and reviewed in [this PR](https://github.com/MetaMask/metamask-mobile/pull/6645). We are now updating it and bringing it into main. 

## Bitrise
- [x] [Flask Bitrise build](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/804ff29b-24bd-4a0a-88b0-4256c57cf7dc) ✅ 
    - Passed. We expect the `deploy android to store` to fail since we need Dan Finlay to help us setup the last part of the pipeline. APKS and IPA are present in the artifacts and iOS has been uploaded to testflight.
  
<img width="1681" alt="Screenshot 2023-11-21 at 6 09 37 PM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/03abdd0e-fac9-49cc-86ac-ecedf0a92dfc">

- [x] [QA Bitrise build](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/d23c35ca-dae0-4d84-9c06-5daa6f7b352c) ✅ 
    - Passed with no issues
- [x] [Release](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/17f0e767-3ab9-42b8-97e8-4116929648bc)
    - Passed with no issues ✅ 
- [x] [E2E Smoke](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/56ff9984-d1a1-45a6-964a-e3054867a9d8) ✅ 
    - Passed with no issues.
- [x] [E2E regression test](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a388972b-c57b-4f3d-b341-0c61cf562c1d) ✅ 
    - Passed with no issues

# Testflight

### Main build
- I created a test production release to verify that these changes `do not effect production releases`. You can test this out on Testflight `v7.12.1 1213`.

https://github.com/MetaMask/metamask-mobile/assets/22918444/4f00d4d0-0d64-47aa-b325-d19e583d1bd1



### Flask Build
- `version 0.0.4 1212`
- This build is available for testing through testflight
- the app is correctly built and everything is working
- the proper code fencing (flask) has been applied

https://github.com/MetaMask/metamask-mobile/assets/22918444/ce7ac0e0-faff-4be5-94ea-e4962f9a9547



## **Related issues**

Progresses: https://github.com/MetaMask/accounts-planning/issues/143

## **Manual testing steps iOS**

1. checkout this PR and run `yarn setup`
2. run `yarn watch:clean`
3. open the app in Xcode
4. click the drop down menu with the orange fox (pictured below) and select `MetaMask-Flask` as your new variant
5. click build (play button)
6. you can also run flask with `yarn run start:ios:flask`
7. the app should build on the emulator with the purple fox icon and the name `MetaMask Flask`
8. open the app and create a wallet
9. the app should function normally


## **Manual testing steps Android**
1. checkout this PR and run `yarn setup`
2. run `yarn watch:clean`
3. open the app in android studio
4. open the build variants through `View/Tool window/Variants`



<img width="557" alt="Screenshot 2023-11-15 at 6 12 10 PM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/d210eee4-a740-48a1-92ea-dc97fde4ee45">

6. In the build variant tab, open the app variants and select flaskDebug

<img width="439" alt="Screenshot 2023-11-15 at 6 14 22 PM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/3726e874-cf1e-4e99-bba7-041329c1ee5d">


7. now click the green build button
8. you can also run flask with `yarn run start:android:flask`
11. the app should build on the emulator with the purple fox icon and the name `MetaMask Flask`
12. open the app and create a wallet
13. the app should function normally


## **Screenshots/Recordings**



### **Android**

![283299619-024d0e9f-3506-441e-a097-9c0114b351c3](https://github.com/MetaMask/metamask-mobile/assets/22918444/e5d3f79b-20fa-4bb0-8010-3734238bc139)


### **iOS**

![283295309-3705e04a-52f4-4912-a77f-efa6aa6732a4](https://github.com/MetaMask/metamask-mobile/assets/22918444/3b037139-a720-4e10-bccd-db792f09e843)

<img width="405" alt="Screenshot 2023-11-15 at 5 41 33 PM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/5b7a76a9-b67d-4eea-82d7-898b296d2fee">

<img width="1029" alt="Screenshot 2023-11-15 at 5 41 46 PM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/b7a54b9e-b38f-4e63-ba79-62efab5e1bb2">


## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
